<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="QTLseqr performs QTL mapping using Bulk Segregant Analysis of Next Gen Sequencing data. The package is an R implementation of the analyses described in Magwene PM, Willis JH, Kelly JK (2011) The Statistics of Bulk Segregant Analysis Using Next Generation Sequencing. PLOS Computational Biology 7(11): e1002255. doi: 10.1371/journal.pcbi.1002255 and Takagi, H., Abe, A., Yoshida, K., Kosugi, S., Natsume, S., Mitsuoka, C., Uemura, A., Utsushi,H., Tamiru, M., Takuno, S., Innan, H., Cano, L. M., Kamoun, S. and Terauchi, R. (2013), QTL-seq: rapid mapping of quantitative trait loci in rice by whole genome resequencing of DNA from two bulked populations. *Plant J*, 74: 174–183. doi:10.1111/tpj.12105">
<title>QTL mapping using Bulk Segregant Analysis of Next Generation Sequencing data. • QTLseqr</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="QTL mapping using Bulk Segregant Analysis of Next Generation Sequencing data.">
<meta property="og:description" content="QTLseqr performs QTL mapping using Bulk Segregant Analysis of Next Gen Sequencing data. The package is an R implementation of the analyses described in Magwene PM, Willis JH, Kelly JK (2011) The Statistics of Bulk Segregant Analysis Using Next Generation Sequencing. PLOS Computational Biology 7(11): e1002255. doi: 10.1371/journal.pcbi.1002255 and Takagi, H., Abe, A., Yoshida, K., Kosugi, S., Natsume, S., Mitsuoka, C., Uemura, A., Utsushi,H., Tamiru, M., Takuno, S., Innan, H., Cano, L. M., Kamoun, S. and Terauchi, R. (2013), QTL-seq: rapid mapping of quantitative trait loci in rice by whole genome resequencing of DNA from two bulked populations. *Plant J*, 74: 174–183. doi:10.1111/tpj.12105">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">QTLseqr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/term_definitions.html">Definition of Terms</a>
    <a class="dropdown-item" href="articles/Yang_et_al_2013.html">Yang_et_al_2013</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bmansfeld/QTLseqr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="qtlseqr-v0752">QTLseqr v0.7.5.2<a class="anchor" aria-label="anchor" href="#qtlseqr-v0752"></a>
</h1></div>
<p>QTLseqr is an R package for QTL mapping using NGS Bulk Segregant Analysis.</p>
<p>QTLseqr is still under development and is offered with out any guarantee.</p>
<div class="section level3">
<h3 id="for-more-detailed-instructions-please-read-the-vignette-here">
<strong>For more detailed instructions please read the vignette <a href="https://github.com/bmansfeld/QTLseqr/raw/master/vignettes/QTLseqr.pdf" class="external-link">here</a></strong><a class="anchor" aria-label="anchor" href="#for-more-detailed-instructions-please-read-the-vignette-here"></a>
</h3>
</div>
<div class="section level3">
<h3 id="for-updates-read-the-newsmd">For updates read the <a href="https://github.com/bmansfeld/QTLseqr/blob/master/NEWS.md" class="external-link">NEWS.md</a>
<a class="anchor" aria-label="anchor" href="#for-updates-read-the-newsmd"></a>
</h3>
</div>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<!-- You can install and update QTLseqr by using our [drat](http://dirk.eddelbuettel.com/code/drat.html) repository hosted on our github page: -->
<!-- ```{r drat-install, eval = FALSE} -->
<!-- install.packages("QTLseqr", repos = "http://bmansfeld.github.io/drat") -->
<!-- ``` -->
<!-- OR You can install QTLseqr from github with: -->
<p>You can install QTLseqr from github with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install devtools first to download packages from github</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># use devtools to install QTLseqr</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bmansfeld/QTLseqr"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> Apart from regular package dependencies, there are some Bioconductor tools that we use as well, as such you will be prompted to install support for Bioconductor, if you haven’t already. QTLseqr makes use of C++ to make some tasks significantly faster (like counting SNPs). Because of this, in order to install QTLseqr from github you will be required to install some compiling tools (Rtools and Xcode, for Windows and Mac, respectively).</p>
<p><strong>If you use QTLseqr in published research, please cite:</strong></p>
<blockquote>
<p>Mansfeld B.N. and Grumet R, QTLseqr: An R package for bulk segregant analysis with next-generation sequencing <em>The Plant Genome</em> <a href="https://dl.sciencesocieties.org/publications/tpg/abstracts/11/2/180006" class="external-link">doi:10.3835/plantgenome2018.01.0006</a></p>
</blockquote>
<p>We also recommend citing the paper for the corresponding method you work with.</p>
<p>QTL-seq method:</p>
<blockquote>
<p>Takagi, H., Abe, A., Yoshida, K., Kosugi, S., Natsume, S., Mitsuoka, C., Uemura, A., Utsushi, H., Tamiru, M., Takuno, S., Innan, H., Cano, L. M., Kamoun, S. and Terauchi, R. (2013), QTL-seq: rapid mapping of quantitative trait loci in rice by whole genome resequencing of DNA from two bulked populations. <em>Plant J</em>, 74: 174–183. <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/tpj.12105" class="external-link">doi:10.1111/tpj.12105</a></p>
</blockquote>
<p>G prime method:</p>
<blockquote>
<p>Magwene PM, Willis JH, Kelly JK (2011) The Statistics of Bulk Segregant Analysis Using Next Generation Sequencing. <em>PLOS Computational Biology</em> 7(11): e1002255. <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002255" class="external-link">doi.org/10.1371/journal.pcbi.1002255</a></p>
</blockquote>
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>Next Generation Sequencing Bulk Segregant Analysis (NGS-BSA) is efficient in detecting quantitative trait loci (QTL). Despite the popularity of NGS-BSA and the R statistical platform, no R packages are currently available for NGS-BSA. We present QTLseqr, an R package for NGS-BSA that identifies QTL using two statistical approaches: QTL-seq and G’. These approaches use a simulation method and a tricube smoothed G statistic, respectively, to identify and assess statistical significance of QTL. QTLseqr, can import and filter SNP data, calculate SNP distributions, relative allele frequencies, G’ values, and log10(p-values), enabling identification and plotting of QTL.</p>
</div>
</div>
<div class="section level1">
<h1 id="examples">Examples:<a class="anchor" aria-label="anchor" href="#examples"></a>
</h1>
<div class="section level2">
<h2 id="example-figure">Example figure<a class="anchor" aria-label="anchor" href="#example-figure"></a>
</h2>
<div class="float">
<img src="https://github.com/bmansfeld/QTLseqr/raw/master/all_plots.png%20%22Example%20figure%22" alt="Example figure"><div class="figcaption">Example figure</div>
</div>
<p>###<strong>For more detailed instructions please read the vignette <a href="https://github.com/bmansfeld/QTLseqr/raw/master/vignettes/QTLseqr.pdf" class="external-link">here</a></strong></p>
<p>This is a basic example which shows you how to import and analyze NGS-BSA data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#load the package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/bmansfeld/QTLseqr/" class="external-link">"QTLseqr"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Set sample and file names</span></span>
<span><span class="va">population_2</span> <span class="op">&lt;-</span> <span class="st">"SRR834931"</span></span>
<span><span class="va">population_1</span> <span class="op">&lt;-</span> <span class="st">"SRR834927"</span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="st">"SNPs_from_GATK.table"</span></span>
<span></span>
<span><span class="co">#Choose which chromosomes will be included in the analysis (i.e. exclude smaller contigs)</span></span>
<span><span class="va">Chroms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Chr"</span>, <span class="fl">12</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Import SNP data from file</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">importFromGATK</span><span class="op">(</span></span>
<span>        file <span class="op">=</span> <span class="va">file</span>,</span>
<span>        population_2 <span class="op">=</span> <span class="va">population_2</span>,</span>
<span>        population_1 <span class="op">=</span> <span class="va">population_1</span>,</span>
<span>        chromList <span class="op">=</span> <span class="va">Chroms</span></span>
<span>     <span class="op">)</span></span>
<span></span>
<span><span class="co">#Filter SNPs based on some criteria</span></span>
<span><span class="va">df_filt</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">filterSNPs</span><span class="op">(</span></span>
<span>        SNPset <span class="op">=</span> <span class="va">df</span>,</span>
<span>        refAlleleFreq <span class="op">=</span> <span class="fl">0.20</span>,</span>
<span>        minTotalDepth <span class="op">=</span> <span class="fl">100</span>,</span>
<span>        maxTotalDepth <span class="op">=</span> <span class="fl">400</span>,</span>
<span>        minSampleDepth <span class="op">=</span> <span class="fl">40</span>,</span>
<span>        minGQ <span class="op">=</span> <span class="fl">99</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#Run G' analysis</span></span>
<span><span class="va">df_filt</span> <span class="op">&lt;-</span> <span class="fu">runGprimeAnalysis</span><span class="op">(</span></span>
<span>    SNPset <span class="op">=</span> <span class="va">df_filt</span>,</span>
<span>    windowSize <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>    outlierFilter <span class="op">=</span> <span class="st">"deltaSNP"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Run QTLseq analysis</span></span>
<span><span class="va">df_filt</span> <span class="op">&lt;-</span> <span class="fu">runQTLseqAnalysis</span><span class="op">(</span></span>
<span>    SNPset <span class="op">=</span> <span class="va">df_filt</span>,</span>
<span>    windowSize <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>    popStruc <span class="op">=</span> <span class="st">"F2"</span>,</span>
<span>    bulkSize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">25</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    replications <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    intervals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">95</span>, <span class="fl">99</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Plot</span></span>
<span><span class="fu">plotQTLStats</span><span class="op">(</span>SNPset <span class="op">=</span> <span class="va">df_filt</span>, var <span class="op">=</span> <span class="st">"Gprime"</span>, plotThreshold <span class="op">=</span> <span class="cn">TRUE</span>, q <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu">plotQTLStats</span><span class="op">(</span>SNPset <span class="op">=</span> <span class="va">df_filt</span>, var <span class="op">=</span> <span class="st">"deltaSNP"</span>, plotIntervals <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#export summary CSV</span></span>
<span><span class="fu">getQTLTable</span><span class="op">(</span>SNPset <span class="op">=</span> <span class="va">df_filt</span>, alpha <span class="op">=</span> <span class="fl">0.01</span>, export <span class="op">=</span> <span class="cn">TRUE</span>, fileName <span class="op">=</span> <span class="st">"my_BSA_QTL.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bmansfeld/QTLseqr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/bmansfeld/QTLseqr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing QTLseqr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ben Nathan Mansfeld <br><small class="roles"> Author, maintainer </small>  </li>
<li>Chase Mateusiak <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-2890-4242" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ben Nathan Mansfeld, Chase Mateusiak.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
